### PR TITLE
Issue #17659: set tick color and tick labelcolor independently from rcParams

### PR DESCRIPTION
**Pull request by rbnvrw**
## PR Summary

This PR implements the behaviour as I requested in issue #17659: it allows one to set the tick color and the tick labelcolor independently from rcParams or a matplotlib style file.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] New features are documented, with examples if plot related: **I have not added any documentation, let me know if I should**
- [x] Documentation is sphinx and numpydoc compliant: **I have not added any documentation, let me know if I should**
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way: **not applicable, the change is backwards compatible**

### CODE REVIEW

File: lib/matplotlib/axis.py [hunk 1]
```diff
@@ -120,7 +120,11 @@ def __init__(self, axes, loc, label=None,
         self._base_pad = pad
 
         if labelcolor is None:
-            labelcolor = mpl.rcParams[f"{name}.color"]
+            labelcolor = mpl.rcParams[f"{name}.labelcolor"]
+            if labelcolor is None:
```
Comment (anntzer) [inline lib/matplotlib/axis.py:124]: if you're using validate_color, this can't be none.  possibly you may want to use validate_color_or_inherit.

Comment (rbnvrw) [reply lib/matplotlib/axis.py:124]: Hi @anntzer thank you for your help. I just copied everything from the existing xtick.color into xtick.labelcolor (and xtick.color has the same `is None` check and the `validate_color`). Should that also change? I'm a bit confused.

Comment (anntzer) [reply lib/matplotlib/axis.py:124]: The argument passed in (`color`, `labelcolor`) can be None, but the rcparam cannot... right? (from a quick look)

**Comment by tacaswell** [comment]
Thanks for working on this!

> not applicable, the change is backwards compatible

I do not agree that this is back compatible.  If someone was changing `'xtick.color'` and expecting both the tick and the label to change colors, now the ticks will remain black.  The default value of the new rcparams should probably be "go look up '{name}.color' instead" rather than hard-coded to a color.  In other cases we have used the string "inherit" for this (see `'legend.facecolor'`)

Another behavior change is previously if you passed the `color` kwargs, the `labelcolor` would fallback to the `xtick.color` rcparam, where as now fallback if `xtick.labelcolor` is set to `None`, it falls back to the `color` variable (which may have been passed in and not be equal to the rcparams.

> probably not applicable, it is a small enhancement

But as you point out it is a small enhancement that may make a bunch of people's lives better.  It is important that we highlight those changes so people know about them!

File: lib/matplotlib/axis.py (hunk 1 above)
Comment (rbnvrw) [reply lib/matplotlib/axis.py:124]: Thanks, I understand now! I changed it.

**Comment by rbnvrw** [comment]
Thank you @tacaswell and @anntzer for your help. I've implemented your suggestions. Let me know what you think.

File: lib/matplotlib/axis.py (hunk 1 above)
Comment (tacaswell) [reply lib/matplotlib/axis.py:124]: I think this conditional can still be removed (as it is handled in the next block).

File: doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
+The color of ticks and tick labels can be set independently from rcParams
+-------------------------------------------------------------------------
+
+The color of ticks and tick labels can now be set independently from the
+matplotlib rc file or the rcParams using the xtick or ytick.color and
+.labelcolor parameters.
```
Comment (tacaswell) [inline doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst:6]: We have custom extension so 
```rst
:rc:`xtick.labelcolor`
```
will render nicely in the output, could you use that here?

Can this text include an example of how to use this?  I'm not sure our average user will understand how to use this new feature from this text.

Comment (rbnvrw) [reply doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst]: Of course! I have updated the text. Let me know if it is better.

File: lib/matplotlib/axis.py (hunk 1 above)
Comment (rbnvrw) [reply lib/matplotlib/axis.py]: You're absolutely right, I forgot that it can only be inherit or a color now. Thanks!

**Comment by rbnvrw** [comment]
Thank you for your help @tacaswell, it was fun to contribute to matplotlib. I don't know a lot about the codebase but your help was really insightful!

**Comment by rbnvrw** [comment]
Hi @tacaswell one test has failed but I'm not sure what I should do to improve it.. it seems to check the coverage of the test files themselves or do I misunderstand it?

File: matplotlibrc.template [hunk 3]
```diff
@@ -461,7 +461,8 @@
 #xtick.minor.width:   0.6     # minor tick width in points
 #xtick.major.pad:     3.5     # distance to major tick label in points
 #xtick.minor.pad:     3.4     # distance to the minor tick label in points
-#xtick.color:         black   # color of the tick labels
+#xtick.color:         black   # color of the ticks
+#xtick.labelcolor:    inherit # color of the tick labels
```
Comment (jklymak) [inline matplotlibrc.template:465]:
```suggestion
#xtick.labelcolor:    inherit # color of the tick labels or inherit from xtick.color
```

File: matplotlibrc.template [hunk 4]
```diff
@@ -481,7 +482,8 @@
 #ytick.minor.width:   0.6     # minor tick width in points
 #ytick.major.pad:     3.5     # distance to major tick label in points
 #ytick.minor.pad:     3.4     # distance to the minor tick label in points
-#ytick.color:         black   # color of the tick labels
+#ytick.color:         black   # color of the ticks
+#ytick.labelcolor:    inherit # color of the tick labels
```
Comment (jklymak) [inline matplotlibrc.template:486]:
```suggestion
#ytick.labelcolor:    inherit # color of the tick labels or inherit from ytick.color
```

**Comment by rbnvrw** [comment]
Thank you for your suggestions @jklymak :+1:

File: doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst [hunk 5]
```diff
@@ -0,0 +1,34 @@
+The color of ticks and tick labels can be set independently from rcParams
+-------------------------------------------------------------------------
+
+The color of ticks and tick labels can now be set independently from the
+matplotlib rc file or the rcParams using the :rc:`xtick.color` or 
+:rc:`ytick.color` parameters to set the tick colors and the 
+:rc:`xtick.labelcolor` or :rc:`ytick.labelcolor` parameters to set the tick
```
Comment (timhoffm) [inline doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst:7]: More concise:
~~~
:rc:`xtick.color` used to define the tick color and the label color. The label
color can now be set independently using :rc:`xtick.labelcolor`. It defaults
to "inherit" which will take the value from :rc:`xtick.color`. The same holds
for ``ytick.[label]color``.
~~~

File: doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+   xtick.labelcolor : lightgrey
+   xtick.color      : black
+   ytick.labelcolor : lightgrey
+   ytick.color      : black
+
+
+The default value for the :rc:`xtick.labelcolor` or :rc:`ytick.labelcolor`
+parameters is 'inherit', in which case the value from :rc:`xtick.color` or
+:rc:`ytick.color` is used.
```
Comment (timhoffm) [inline doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst:34]: Can be omitted after the canges above.

File: doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst (hunk 5 above)
Comment (rbnvrw) [reply doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst]: Thank you @timhoffm I've implemented your suggestions.

File: doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst [hunk 7]
```diff
@@ -0,0 +1,31 @@
+The color of ticks and tick labels can be set independently using rcParams
+--------------------------------------------------------------------------
+
+The color of ticks and tick labels can now be set independently from the
```
Comment (timhoffm) [inline doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst:4]: This sentence essentially only repeats the heading and can thus be left out. The only addition (mentioning of rc file) is also covered below.

Comment (rbnvrw) [reply doc/users/next_whats_new/set_colors_ticks_and_ticklabels_rcparams.rst]: OK, I hope it is concise enough now!

**Comment by timhoffm** [comment]
Thanks, and congratulations on your first contribution to Matplotlib! Hope to see you back some time.

**Comment by rbnvrw** [comment]
> Thanks, and congratulations on your first contribution to Matplotlib! Hope to see you back some time.

Great thank you! Me too :+1:

**Pull request merged**